packages/apollo-server-core/src/utils/schemaManager.ts

```diff
@@ -0,0 +1,216 @@
+      options.schemaDerivedDataProvider(options.apiSchema);
+    }
+  }
+
+  /**
+   * Calling start() will:
+   * - Start gateway schema fetching (if a gateway was provided).
+   * - Initialize schema-derived data.
+   * - Synchronously notify onSchemaLoadOrUpdate() listeners of schema load, and
+   *   asynchronously notify them of schema updates.
+   */
+  public async start(): Promise<void> {
+    if (this.modeSpecificState.mode === 'gateway') {
+      const gateway = this.modeSpecificState.gateway;
+      if (gateway.onSchemaLoadOrUpdate) {
+        // Use onSchemaLoadOrUpdate if available, as it reports the core
+        // supergraph SDL and always reports the initial schema load.
+        this.modeSpecificState.unsubscribeFromGateway =
+          gateway.onSchemaLoadOrUpdate((schemaContext) => {
+            this.processSchemaLoadOrUpdateEvent(schemaContext);
+          });
+      } else {
+        this.modeSpecificState.unsubscribeFromGateway = gateway.onSchemaChange(
+          (apiSchema) => {
+            this.processSchemaLoadOrUpdateEvent({ apiSchema });
+          },
+        );
+      }
+
+      const config = await this.modeSpecificState.gateway.load({
+        apollo: this.modeSpecificState.apolloConfig,
+      });
+
+      // Note that for old gateways that have onSchemaChange() and no
+      // onSchemaLoadOrUpdate(), this.schemaDerivedData may not be initialized
+      // during gateway.load() (because old gateways don't notify listeners on
+      // schema load in some cases), so we must initialize it here if needed.
+      if (!this.schemaDerivedData) {
+        this.processSchemaLoadOrUpdateEvent({ apiSchema: config.schema });
+      }
+    } else {
+      this.processSchemaLoadOrUpdateEvent({
+        apiSchema: this.modeSpecificState.apiSchema,
+      });
+    }
+  }
+
+  /**
+   * Registers a listener for schema load/update events. Note that the latest
+   * event is buffered, i.e.
+   * - If registered before start(), then the callback will be first called in
+   *   start(), and then later for updates.
+   * - If registered after start() but before stop(), the callback will be first
+   *   called in this method (for whatever the current schema is), and then
+   *   later for updates.
+   * - If registered after stop(), the callback will never be called.
+   *
+   * For gateways, a core supergraph SDL will be provided to the callback. If
+   * your gateway is too old to provide a core supergraph SDL, this method will
+   * throw.
+   *
+   * @param callback The listener to execute on schema load/updates.
+   */
+  public onSchemaLoadOrUpdate(
+    callback: (schemaContext: GraphQLSchemaContext) => void,
+  ): Unsubscriber {
+    if (
+      this.modeSpecificState.mode === 'gateway' &&
+      !this.modeSpecificState.gateway.onSchemaLoadOrUpdate
+    ) {
+      throw new Error(
+        'Your gateway is too old to register a onSchemaLoadOrUpdate() listener',
```

Comment:
It would be nice if we could get an `@apollo/gateway` version number to upgrade to in here, though it'll take some placeholdering until we get releases coordinated.

Reply:
Yep, agreed it would be better to have a version number here, I've added a `FIXME` to this.